presto-main/src/main/java/com/facebook/presto/operator/PagesSpatialIndexFactory.java

```diff
@@ -22,25 +22,54 @@
 
 import java.util.ArrayList;
 import java.util.List;
+import java.util.function.Supplier;
 
 import static com.google.common.util.concurrent.Futures.immediateFuture;
 import static java.util.Objects.requireNonNull;
 
+/**
+ * This factory is used to pass spatial index built by SpatialIndexBuilderOperator
+ * to SpatialJoinOperators.
+ * <p>
+ * SpatialIndexBuilderOperator creates a spatial index {@link Supplier} and provides
+ * it to this factory by calling {@link #lendPagesSpatialIndex(Supplier)}.
+ * <p>
+ * SpatialJoinOperators call {@link #createPagesSpatialIndex()} to get a Future that
+ * will provide a {@link Supplier} of the spatial index when done. The {@link Supplier}
+ * is used to create separate instances of an index for each SpatialJoinOperators. All
+ * these instances share the index of geometries on the build side, but have their own
+ * instance of an optional extra potentially stateful filter function.
+ * <p>
+ * Multiple instances of the SpatialJoinOperator share an index created by a single
+ * SpatialIndexBuilderOperator. Therefore, SpatialIndexBuilderOperator is responsible
+ * for keeping track of how much memory is used by the index. To do so
+ * SpatialIndexBuilderOperator has to stay active until all the SpatialJoinOperators
+ * have finished.
+ * <p>
+ * This factory keeps track of the number of active SpatialJoinOperators in cooperation
+ * with SpatialJoinOperatorFactory and SpatialJoinOperators. Initially, the number of
+ * active SpatialJoinOperators is set to 1. When SpatialJoinOperator is created, it calls
+ * {@link #createPagesSpatialIndex()} and the counter goes up by 1. When SpatialJoinOperator
+ * finishes, it calls {@link #probeOperatorFinished()} and the counter goes down by 1.
+ * In addition SpatialJoinOperatorFactory calls {@link #noMoreProbeOperators()} when done
+ * creating operators and the counter goes down by 1.
+ * <p>
+ * {@link #lendPagesSpatialIndex(Supplier)} returns a Future that completes once all
+ * the SpatialJoinOperators completed. SpatialIndexBuilderOperator uses that Future
+ * to decide on its own completion.
```

Comment:
This javadoc is very nice documentation of how things are done now. I'm concerned it may become outdated without notice though.